open_spiel/games/dark_hex.cc

```diff
@@ -0,0 +1,267 @@
+// Copyright 2019 DeepMind Technologies Ltd. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "open_spiel/games/dark_hex.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "open_spiel/games/hex.h"
+#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
+#include "open_spiel/spiel_utils.h"
+
+namespace open_spiel {
+namespace dark_hex {
+namespace {
+  
+using hex::kCellStates;
+using hex::kDefaultBoardSize;
+
+using hex::CellState;
+using hex::kMinValueCellState;
+
+using hex::StateToString;
+using hex::PlayerToState;
+
+// Game Facts
+const GameType kGameType{
+  /*short_name=*/"dark_hex",
+  /*long_name=*/"Dark Hex",
+  GameType::Dynamics::kSequential,
+  GameType::ChanceMode::kDeterministic,
+  GameType::Information::kImperfectInformation,
+  GameType::Utility::kZeroSum,
+  GameType::RewardModel::kTerminal,
+  /*max_num_players=*/2,
+  /*min_num_players=*/2,
+  /*provides_information_state_string=*/true,
+  /*provides_information_state_tensor=*/true,
+  /*provides_observation_string=*/true,
+  /*provides_observation_tensor=*/true,
+  /*parameter_specification=*/
+  {
+    {"obstype", GameParameter(static_cast<std::string>(kDefaultObsType))},
+    {"board_size", GameParameter(hex::kDefaultBoardSize)},
+  }
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params){
+  return std::shared_ptr<const Game>(
+    new DarkHexGame(params)
+  );
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+} // namespace
+
+DarkHexState::DarkHexState(std::shared_ptr<const Game> game, int board_size,
+                                  ObservationType obs_type)
+    : State(game), state_(game, board_size), board_size_(board_size), 
+                          obs_type_(obs_type) {
+  black_view_.resize(board_size * board_size, CellState::kEmpty);
+  white_view_.resize(board_size * board_size, CellState::kEmpty);
+
+  kNumCells = board_size_ * board_size_;
+  kBitsPerAction = kNumCells + 1;
+  kLongestSequence = kNumCells * 2 - 1;
+}
+
+void DarkHexState::DoApplyAction(Action move) {
+  Player cur_player = CurrentPlayer(); //current player
+  auto& cur_view = (cur_player == Player(0) ? black_view_ : white_view_);
+
+  // Either occupied or not
+  if (state_.BoardAt(move) == CellState::kEmpty) {
+    state_.ApplyAction(move);
+  } else {}
```

Comment:
Why have an empty else here?

Reply:
https://github.com/deepmind/open_spiel/blob/283ab4c7725ba753cc2b85581c398471d05a5574/open_spiel/games/phantom_ttt.cc#L81 I got it from here

Reply:
Ok let's remove it. You can also remove the one from Phantom Tic-Tac-Toe as well if you want. (If not, I will separately.